Modix.Services/Tags/TagService.cs

```diff
@@ -15,249 +17,129 @@
-        /// Determines whether a tag with the given name exists within the given guild.
-        /// </summary>
-        /// <param name="guildId">The Discord snowflake ID value of the guild to which the tag belongs.</param>
-        /// <param name="name">The name that is used to invoke the tag.</param>
-        /// <returns>True if the tag exists, false if not</returns>
         Task<bool> TagExistsAsync(ulong guildId, string name);
     }
 
-    /// <inheritdoc />
     internal class TagService : ITagService
     {
+        private readonly IDiscordClient _discordClient;
+        private readonly IAuthorizationService _authorizationService;
+        private readonly IUserService _userService;
+        private readonly IDesignatedRoleMappingRepository _designatedRoleMappingRepository;
+        private readonly ModixContext _modixContext;
+
         private static readonly Regex _tagNameRegex = new Regex(@"^\S+\b$");
-        
-        /// <summary>
-        /// Constructs a new <see cref="TagService"/> with the supplied dependencies.
-        /// </summary>
+
         public TagService(
             IDiscordClient discordClient,
             IAuthorizationService authorizationService,
-            ITagRepository tagRepository,
             IUserService userService,
-            IDesignatedRoleMappingRepository designatedRoleMappingRepository)
+            IDesignatedRoleMappingRepository designatedRoleMappingRepository,
+            ModixContext modixContext)
         {
-            DiscordClient = discordClient;
-            AuthorizationService = authorizationService;
-            TagRepository = tagRepository;
-            UserService = userService;
-            DesignatedRoleMappingRepository = designatedRoleMappingRepository;
+            _discordClient = discordClient;
+            _authorizationService = authorizationService;
+            _userService = userService;
+            _modixContext = modixContext;
+            _designatedRoleMappingRepository = designatedRoleMappingRepository;
         }
 
-        /// <inheritdoc />
         public async Task CreateTagAsync(ulong guildId, ulong creatorId, string name, string content)
         {
-            AuthorizationService.RequireClaims(AuthorizationClaim.CreateTag);
+            _authorizationService.RequireClaims(AuthorizationClaim.CreateTag);
 
             if (string.IsNullOrWhiteSpace(name))
                 throw new ArgumentException("The tag name cannot be blank or whitespace.", nameof(name));
 
             if (string.IsNullOrWhiteSpace(content))
                 throw new ArgumentException("The tag content cannot be blank or whitespace.", nameof(content));
-                
+
             if (!_tagNameRegex.IsMatch(name))
                 throw new ArgumentException("The tag name cannot have punctuation at the end.", nameof(name));
 
             name = name.Trim().ToLower();
 
-            using (var transaction = await TagRepository.BeginMaintainTransactionAsync())
```

Comment:
(Nox) EF Core acquires its transaction on calls to SaveChanges (and is thus atomic for all changes thusly saved), or explicit calls to `context.Database.BeginTransaction()`. 
Concurrent request can become an issue if you rely on the ChangeTracker - that is, if both you and I retrieve the same entity, and I save the entity, **then** you save the entity, EF will return a DbConcurrencyException (or the like) which you have to manually resolve in-code (decide which entry wins).

Reply:
From what I can ascertain here, the transactions existed because of the multiple invocations of `SaveChanges` - which, by default will lead EF to wrap any changes in a transaction. Because there were multiple saves, it could fail which would require the rollback.

As for the atomicity - that _could_ be an issue. This naively will query out the tag immediately and if a tag is used X or more times, at the same time, it will increment from whatever value was queried at the time. I consider this more of `async lock` if that makes any sense.

Reply:
The transaction system we need to move to is `System.Transactions` and gut all the `IRepositoryTransaction` stuff. It solves the same problems as my crap transactions system, and solves them much better. The only thing it doesn't have is mockability, which is easily done with a small, custom `ITransactionScopeFactory` implementation that's injected into services